OutOfSchool/OutOfSchool.Redis/MultiLayerCache.cs

```diff
@@ -0,0 +1,125 @@
+        this.memoryCacheConfig = memoryCacheConfig.Value;
+    }
+
+    public Task ClearCacheAsync(string key)
+    {
+        return ExecuteMultipleAsync(() =>
+        {
+            cacheLock.EnterWriteLock();
+            try
+            {
+                _memoryCache.Remove(key);
+                _cacheService.ClearCacheAsync(key);
+            }
+            finally
+            {
+                cacheLock.ExitWriteLock();
+            }
+        });
+    }
+
+    public async Task<T> GetOrAddAsync<T>(string key, Func<Task<T>> newValueFactory, TimeSpan? absoluteExpirationRelativeToNowInterval = null, TimeSpan? slidingExpirationInterval = null)
+    {
+        T returnValue = default;
+
+        await ExecuteMultipleAsync(() =>
+        {
+            cacheLock.EnterReadLock();
+            
+            try
+            {
+                if (_memoryCache.TryGetValue(key, out T value))
+                {
+                    returnValue = value;
+                }
+            }
+            finally
+            {
+                cacheLock.ExitReadLock();
+            }
+        });
+
+        if (EqualityComparer<T>.Default.Equals(returnValue, default))
+        {
+            returnValue = await _cacheService.GetOrAddAsync(key, newValueFactory, absoluteExpirationRelativeToNowInterval, slidingExpirationInterval);
+            await SetAsync(key, returnValue, absoluteExpirationRelativeToNowInterval, slidingExpirationInterval);
+        }
+
+        return returnValue;
+    }
+
+    public Task RefreshAsync(string key)
+    {
+        return ExecuteMultipleAsync(() =>
+        {
+            cacheLock.EnterWriteLock();
+            try
+            {
+                _cacheService.RefreshAsync(key);
+            }
+            finally
+            {
+                cacheLock.ExitWriteLock();
+            }
+        });
+    }
+
+    public Task SetAsync<T>(string key, T value, TimeSpan? absoluteExpirationRelativeToNowInterval = null, TimeSpan? slidingExpirationInterval = null)
+    {
+        return ExecuteMultipleAsync(() =>
+        {
+            cacheLock.EnterWriteLock();
+
+            try
+            {
+                var cacheEntryOptions = new MemoryCacheEntryOptions()
+                {
+                    SlidingExpiration = slidingExpirationInterval ?? memoryCacheConfig.SlidingExpirationInterval,
+                    AbsoluteExpirationRelativeToNow = absoluteExpirationRelativeToNowInterval ?? memoryCacheConfig.AbsoluteExpirationRelativeToNowInterval,
+                    Priority = CacheItemPriority.Normal,
+                    Size = 1024,
+                };
+
+                _memoryCache.Set(key, value, cacheEntryOptions);
+            }
+            finally
+            {
+                cacheLock.ExitWriteLock();
+            }
+        });
+    }
+
+    public void Dispose()
+    {
+        cacheLock?.Dispose();
+    }
+
+    private Task ExecuteMultipleAsync(Action operation)
```

Comment:
Please provide more details why this method is named `ExecuteMultipleAsync` and why we need it at all?
As I can see from the usage it's only purpose is to make synchronous methods async. Why don't have it stated clearly like
```
public void ClearCache(string key) // <--- here the "cache logic" resides
{
    cacheLock.EnterWriteLock();
    try
    {
        _memoryCache.Remove(key);
        _cacheService.ClearCacheAsync(key);
    }
    finally
    {
        cacheLock.ExitWriteLock();
    }
}

public Task ClearCacheAsync(string key) // <--- here the "async logic" resides
  => Task.Run(() => ClearCache(key));
```  
Thus you have `cache` logic separated from `async` flavour.

Reply:
The question is still open.
The suggested code change is no longer viable, as I found in other comment async call to `_cacheService.ClearCacheAsync(key);`.

Reply:
I can`t get you an answer about naming. It smth with my deep night minds. I renamed it to ExecuteCacheMethod.

Reply:
There is method that does only one action - calls `Task.Run`. 
1. Why it is named `ExecuteMultipleAsync`/`ExecuteCacheMethod`? What code inside this method is related to `Multiple` or to `Cache`? 
2. Why this method is needed at all? What if inline it?

Reply:
I deleted this method